File: research/training/signal_neural/metrics.py
"""Evaluation metrics — byte-for-byte mirror of `cohort-eval.ts`.

Same definitions as the classical TS cohort so a "neural vs classical" comparison
is apples-to-apples:
  - balanced accuracy = mean per-class recall over PRESENT classes (chance = 1/K,
    immune to class imbalance — a skewed prior cannot inflate it),
  - macro-F1 over present classes,
  - ECE = 10-bin top-probability expected calibration error,
  - confusion matrix + per-class precision/recall/F1.
"""
from __future__ import annotations

from typing import Sequence

import numpy as np


def present_labels(y_true: Sequence[str], labels: Sequence[str]) -> list[str]:
    s = set(y_true)
    return [l for l in labels if l in s]


def balanced_accuracy(y_true: Sequence[str], y_pred: Sequence[str], labels: Sequence[str]) -> float:
    total = 0.0
    used = 0
    for l in labels:
        tp = 0
        support = 0
        for t, p in zip(y_true, y_pred):
            if t == l:
                support += 1
                if p == l:
                    tp += 1
        if support > 0:
            total += tp / support
            used += 1
    return total / used if used > 0 else 0.0


def accuracy(y_true: Sequence[str], y_pred: Sequence[str]) -> float:
    if not y_true:
        return 0.0
    return sum(1 for t, p in zip(y_true, y_pred) if t == p) / len(y_true)


def per_class(y_true: Sequence[str], y_pred: Sequence[str], labels: Sequence[str]) -> list[dict]:
    out = []
    for label in labels:
        tp = fp = fn = support = 0
        for t, p in zip(y_true, y_pred):
            tt = t == label
            pp = p == label
            if tt:
                support += 1
            if tt and pp:
                tp += 1
            elif (not tt) and pp:
                fp += 1
            elif tt and (not pp):
                fn += 1
        precision = tp / (tp + fp) if (tp + fp) > 0 else 0.0
        recall = tp / (tp + fn) if (tp + fn) > 0 else 0.0
        f1 = (2 * precision * recall) / (precision + recall) if (precision + recall) > 0 else 0.0
        out.append({"label": label, "support": support, "precision": precision, "recall": recall, "f1": f1})
    return out


def macro_f1(y_true: Sequence[str], y_pred: Sequence[str], labels: Sequence[str]) -> float:
    pc = per_class(y_true, y_pred, labels)
    return sum(c["f1"] for c in pc) / len(pc) if pc else 0.0


def confusion(y_true: Sequence[str], y_pred: Sequence[str], labels: Sequence[str]) -> dict:
    idx = {l: i for i, l in enumerate(labels)}
    m = [[0 for _ in labels] for _ in labels]
    for t, p in zip(y_true, y_pred):
        if t in idx and p in idx:
            m[idx[t]][idx[p]] += 1
    return {"labels": list(labels), "matrix": m}


def expected_calibration_error(y_true: Sequence[str], y_pred: Sequence[str], p_top: Sequence[float], bins: int = 10) -> float:
    bin_acc = [0.0] * bins
    bin_conf = [0.0] * bins
    bin_n = [0] * bins
    for t, p, pt in zip(y_true, y_pred, p_top):
        b = min(bins - 1, max(0, int(pt * bins)))
        bin_n[b] += 1
        bin_conf[b] += pt
        if t == p:
            bin_acc[b] += 1
    n = len(p_top) or 1
    ece = 0.0
    for b in range(bins):
        if bin_n[b] == 0:
            continue
        ece += (bin_n[b] / n) * abs(bin_acc[b] / bin_n[b] - bin_conf[b] / bin_n[b])
    return ece


def majority_class_accuracy(y_true: Sequence[str]) -> float:
    if not y_true:
        return 0.0
    vals, counts = np.unique(np.array(y_true), return_counts=True)
    return float(counts.max() / len(y_true))


def full_metrics(y_true: Sequence[str], y_pred: Sequence[str], p_top: Sequence[float], labels: Sequence[str]) -> dict:
    present = present_labels(y_true, labels)
    pc = per_class(y_true, y_pred, present)
    return {
        "n": len(y_true),
        "num_classes": len(present),
        "accuracy": accuracy(y_true, y_pred),
        "balanced_accuracy": balanced_accuracy(y_true, y_pred, present),
        "macro_f1": (sum(c["f1"] for c in pc) / len(pc)) if pc else 0.0,
        "ece": expected_calibration_error(y_true, y_pred, p_top),
        "per_class": pc,
        "confusion": confusion(y_true, y_pred, present),
        "chance": {
            "majority_class_accuracy": majority_class_accuracy(y_true),
            "balanced_chance": 1.0 / max(1, len(present)),
        },
    }

File: packages/shared-types/src/metrics.ts
/**
 * Classification evaluation metrics — the single source for the honest offline
 * protocol shared by signal-lab's `evaluate` (fusion-label target) and `cohort-eval`
 * (generic target) harnesses. Pure, dependency-free, browser-safe: top-1 accuracy,
 * balanced accuracy, per-class precision/recall/F1, confusion matrix, top-class ECE,
 * and grouped-CV fold assignment.
 *
 * Balanced accuracy is the honest promotion metric: its chance level is a flat
 * 1/numClasses regardless of class imbalance, so an accuracy claim cannot hide behind
 * a skewed prior (raw accuracy can).
 */

/** Per-class precision / recall / F1 with support count. */
export interface ClassMetric {
  readonly label: string;
  readonly support: number;
  readonly precision: number;
  readonly recall: number;
  readonly f1: number;
}

/** Plain top-1 accuracy. Returns 0 for empty input. */
export function accuracyOf(yTrue: readonly string[], yPred: readonly string[]): number {
  if (yTrue.length === 0) return 0;
  let correct = 0;
  for (let i = 0; i < yTrue.length; i++) if (yTrue[i] === yPred[i]) correct++;
  return correct / yTrue.length;
}

/**
 * Balanced accuracy = mean per-class recall over classes WITH support. Chance level
 * is a flat 1/numClasses regardless of imbalance, so it can't be gamed by predicting
 * the majority class. Returns 0 when no class has support.
 */
export function balancedAccuracy(yTrue: readonly string[], yPred: readonly string[], labels: readonly string[]): number {
  let sum = 0;
  let used = 0;
  for (const l of labels) {
    let tp = 0;
    let support = 0;
    for (let i = 0; i < yTrue.length; i++) {
      if (yTrue[i] === l) {
        support++;
        if (yPred[i] === l) tp++;
      }
    }
    if (support > 0) {
      sum += tp / support;
      used++;
    }
  }
  return used > 0 ? sum / used : 0;
}

/** Per-class precision / recall / F1 / support for each label. */
export function perClassMetrics(yTrue: readonly string[], yPred: readonly string[], labels: readonly string[]): ClassMetric[] {
  return labels.map((label) => {
    let tp = 0;
    let fp = 0;
    let fn = 0;
    let support = 0;
    for (let i = 0; i < yTrue.length; i++) {
      const t = yTrue[i] === label;
      const p = yPred[i] === label;
      if (t) support++;
      if (t && p) tp++;
      else if (!t && p) fp++;
      else if (t && !p) fn++;
    }
    const precision = tp + fp > 0 ? tp / (tp + fp) : 0;
    const recall = tp + fn > 0 ? tp / (tp + fn) : 0;
    const f1 = precision + recall > 0 ? (2 * precision * recall) / (precision + recall) : 0;
    return { label, support, precision, recall, f1 };
  });
}

/** Row-true × col-pred confusion matrix over `labels`. */
export function confusionMatrix(
  yTrue: readonly string[],
  yPred: readonly string[],
  labels: readonly string[],
): { labels: string[]; matrix: number[][] } {
  const index = new Map(labels.map((l, i) => [l, i]));
  const matrix = labels.map(() => labels.map(() => 0));
  for (let i = 0; i < yTrue.length; i++) {
    const r = index.get(yTrue[i]!);
    const c = index.get(yPred[i]!);
    if (r !== undefined && c !== undefined) matrix[r]![c]!++;
  }
  return { labels: [...labels], matrix };
}

/**
 * Top-class Expected Calibration Error over `bins` equal-width confidence bins
 * (default 10). The `Math.max(0, …)` bin-floor guard keeps a non-positive top
 * probability from indexing bin −1; for a valid probability in [0, 1] the floor is
 * already ≥ 0, so real values are unchanged (it is purely defensive).
 */
export function expectedCalibrationError(
  yTrue: readonly string[],
  yPred: readonly string[],
  pTop: readonly number[],
  bins = 10,
): number {
  const binAcc = new Array(bins).fill(0);
  const binConf = new Array(bins).fill(0);
  const binN = new Array(bins).fill(0);
  for (let i = 0; i < pTop.length; i++) {
    const b = Math.min(bins - 1, Math.max(0, Math.floor(pTop[i]! * bins)));
    binN[b]++;
    binConf[b] += pTop[i]!;
    if (yTrue[i] === yPred[i]) binAcc[b]++;
  }
  let ece = 0;
  const n = pTop.length || 1;
  for (let b = 0; b < bins; b++) {
    if (binN[b] === 0) continue;
    ece += (binN[b] / n) * Math.abs(binAcc[b] / binN[b] - binConf[b] / binN[b]);
  }
  return ece;
}

/**
 * Assign each sample to one of `folds` CV folds by its `group`, round-robin over
 * sorted group ids so a group's rows never split across folds (no group leakage).
 */
export function groupFolds<T extends { readonly group: string }>(samples: readonly T[], folds: number): number[] {
  const groups = Array.from(new Set(samples.map((s) => s.group))).sort();
  const groupFold = new Map<string, number>();
  groups.forEach((g, i) => groupFold.set(g, i % folds));
  return samples.map((s) => groupFold.get(s.group)!);
}
